platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformWorkspaceIdentity.java

```diff
@@ -17,8 +17,88 @@
 package org.gradle.api.internal.artifacts.transform;
 
 import org.gradle.internal.execution.UnitOfWork;
+import org.gradle.internal.fingerprint.CurrentFileCollectionFingerprint;
+import org.gradle.internal.hash.HashCode;
+import org.gradle.internal.hash.Hasher;
+import org.gradle.internal.hash.Hashing;
 import org.gradle.internal.snapshot.ValueSnapshot;
 
-public interface TransformWorkspaceIdentity extends UnitOfWork.Identity {
-    ValueSnapshot getSecondaryInputsSnapshot();
+class TransformWorkspaceIdentity implements UnitOfWork.Identity {
+    private final ValueSnapshot secondaryInputsSnapshot;
+    private final String uniqueId;
+
+    private TransformWorkspaceIdentity(ValueSnapshot secondaryInputsSnapshot, HashCode uniqueId) {
+        this.uniqueId = uniqueId.toString();
+        this.secondaryInputsSnapshot = secondaryInputsSnapshot;
+    }
+
+    public ValueSnapshot getSecondaryInputsSnapshot() {
+        return secondaryInputsSnapshot;
+    }
+
+    @Override
+    public String getUniqueId() {
+        return uniqueId;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        TransformWorkspaceIdentity that = (TransformWorkspaceIdentity) o;
+
+        return uniqueId.equals(that.uniqueId);
+    }
+
+    @Override
+    public int hashCode() {
+        return uniqueId.hashCode();
+    }
+
+    public static TransformWorkspaceIdentity createMutable(
+        String normalizedInputArtifactPath,
+        String producerBuildTreePath,
+        ValueSnapshot secondaryInputsSnapshot,
+        HashCode dependenciesHash
+    ) {
+        Hasher hasher = Hashing.newHasher();
+        hasher.putString(normalizedInputArtifactPath);
+        hasher.putString(producerBuildTreePath);
+        secondaryInputsSnapshot.appendToHasher(hasher);
+        hasher.putHash(dependenciesHash);
+        return new TransformWorkspaceIdentity(secondaryInputsSnapshot, hasher.hash());
```

Comment:
🤔 I am surprised you can calculate the unique id up front now, given that we had the different subclasses with specialized identities before so we'd avoid calculating them when doing the locking on the loading cache. Are you sure this doesn't introduce any performance regressions?

Reply:
Yeah, I was also somewhat surprised. But ran the performance tests and they show no regression, see link above.